polymetis/polymetis/python/polysim/envs/habitat_manipulator.py

```diff
@@ -0,0 +1,286 @@
+        self.habitat_cfg = make_configuration(habitat_dir, glb_path=habitat_scene_path)
+        self.sim = habitat_sim.Simulator(self.habitat_cfg)
+        place_agent(self.sim)
+
+        # Load robot
+        self.robot = (
+            self.sim.get_articulated_object_manager().add_articulated_object_from_urdf(
+                self.robot_description_path, fixed_base=True
+            )
+        )
+        assert self.robot is not None
+        place_robot_from_agent(self.sim, self.robot)
+
+        self.robot.auto_clamp_joint_limits = True
+
+        # Set correct joint damping values
+        for motor_id in range(self.n_dofs):
+            joint_motor_settings = habitat_sim.physics.JointMotorSettings(
+                0.0,  # position_target
+                0.0,  # position_gain
+                0.0,  # velocity_target
+                joint_damping,  # velocity_gain
+                0.0,  # max_impulse
+            )
+            self.robot.update_joint_motor(motor_id, joint_motor_settings)
+
+        self.reset()
+
+        self.prev_torques_commanded = np.zeros(self.n_dofs)
+        self.prev_torques_applied = np.zeros(self.n_dofs)
+        self.prev_torques_measured = np.zeros(self.n_dofs)
+        self.prev_torques_external = np.zeros(self.n_dofs)
+
+    def reset(self, joint_pos: List[float] = None, joint_vel: List[float] = None):
+        """Reset joint positions / velocities to given values (0s by default)"""
+        if joint_pos is None:
+            joint_pos = OmegaConf.to_container(self.robot_model_cfg.rest_pose)
+        if joint_vel is None:
+            joint_vel = [0.0] * self.n_dofs
+
+        self.robot.joint_positions = joint_pos
+        self.robot.joint_velocities = joint_vel
+
+    def get_num_dofs(self) -> int:
+        """Return number of degrees of freedom."""
+        return self.n_dofs
+
+    def get_current_joint_pos_vel(self) -> Tuple[List[float], List[float]]:
+        """Return tuple of joint positions and velocities, both of size `num_dofs`."""
+        return self.robot.joint_positions, self.robot.joint_velocities
+
+    def get_current_joint_torques(
+        self,
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """Returns torques: [inputted, clipped, added with gravity compensation, and measured externally]"""
+        return (
+            self.prev_torques_commanded,
+            self.prev_torques_applied,
+            self.prev_torques_measured,
+            self.prev_torques_external,
+        )
+
+    def compute_grav_comp(self, pos: List[float], vel: List[float]) -> np.ndarray:
+        """Computes gravity compensation torques using Pinocchio robot model."""
+        return (
+            self.robot_model.inverse_dynamics(
+                torch.tensor(pos),
+                torch.tensor(vel),
+                torch.tensor([0 for _ in range(self.n_dofs)]),
+            )
+            .detach()
+            .cpu()
+            .numpy()
+        )
```

Comment:
By having the current velocity as input you are also compensating for the Coriolis term in addition to gravitational forces. This differs from the [implementation in bullet_manipulator.py](https://github.com/facebookresearch/droidlet/blob/59cc99714703831e5c8c10dc5c4d9ac2c633c4a2/polymetis/polymetis/python/polysim/envs/bullet_manipulator.py#L161), which only compensates for gravity.
Is this intentional due to discrepancies between Pybullet and Habitat?

Reply:
Nope, just fixed, thanks!

Reply:
Nope, just fixed, thanks! Still doesn't do the right thing if we just input gravcomp torques, though; requires some more digging.